{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The IMGT/GENE-DB FASTA header contains 15 fields separated by '|':\n",
    "1. IMGT/LIGM-DB accession number(s)\n",
    "2. IMGT gene and allele name\n",
    "3. species (may be followed by an \"_\" and the name of the strain, breed or isolate, if defined)\n",
    "4. IMGT gene and allele functionality\n",
    "5. exon(s), region name(s), or extracted label(s)\n",
    "6. start and end positions in the IMGT/LIGM-DB accession number(s)\n",
    "7. number of nucleotides in the IMGT/LIGM-DB accession number(s)\n",
    "8. codon start, or 'NR' (not relevant) for non coding labels\n",
    "9. +n: number of nucleotides (nt) added in 5' compared to the corresponding label extracted from IMGT/LIGM-DB\n",
    "10. +n or -n: number of nucleotides (nt) added or removed in 3' compared to the corresponding label extracted from IMGT/LIGM-DB\n",
    "11. +n, -n, and/or nS: number of added, deleted, and/or substituted nucleotides to correct sequencing errors, or 'not corrected' if non corrected sequencing errors\n",
    "12. number of amino acids (AA): this field indicates that the sequence is in amino acids\n",
    "13. number of characters in the sequence: nt (or AA)+IMGT gaps=total\n",
    "14. partial (if it is)\n",
    "15. reverse complementary (if it is)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "columns = [\n",
    "    \"accession number\",\n",
    "    \"name\",\n",
    "    \"species\",\n",
    "    \"functionality\",\n",
    "    \"exon/name/label\",\n",
    "    \"start end positions\",\n",
    "    \"number of nucleotides\",\n",
    "    \"codon start\",\n",
    "    \"+n\",\n",
    "    \"+/-n\",\n",
    "    \"error corrections\",\n",
    "    \"number of AA\",\n",
    "    \"number of char\",\n",
    "    \"partial\",\n",
    "    \"reverse complementary\",\n",
    "    \"FASTA\"]\n",
    "df = pd.DataFrame(columns=columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16\n"
     ]
    }
   ],
   "source": [
    "print(len(columns))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('IMGTGENEDB-ReferenceSequences.fasta-AA-WithoutGaps-F+ORF+inframeP', 'r') as file:\n",
    "    row = ''\n",
    "    while True:\n",
    "        line = file.readline().replace('\\n', '')\n",
    "        if line:\n",
    "            if line.startswith('>'):\n",
    "                if row:\n",
    "                    df.loc[len(df)] = row.split('|')\n",
    "                row = line[1:]\n",
    "            else:\n",
    "                row += line\n",
    "        else:\n",
    "            df.loc[len(df)] = row.split('|')\n",
    "            break       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession number</th>\n",
       "      <th>name</th>\n",
       "      <th>species</th>\n",
       "      <th>functionality</th>\n",
       "      <th>exon/name/label</th>\n",
       "      <th>start end positions</th>\n",
       "      <th>number of nucleotides</th>\n",
       "      <th>codon start</th>\n",
       "      <th>+n</th>\n",
       "      <th>+/-n</th>\n",
       "      <th>error corrections</th>\n",
       "      <th>number of AA</th>\n",
       "      <th>number of char</th>\n",
       "      <th>partial</th>\n",
       "      <th>reverse complementary</th>\n",
       "      <th>FASTA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KT723008</td>\n",
       "      <td>IGHA*01</td>\n",
       "      <td>Bos taurus_Holstein</td>\n",
       "      <td>F</td>\n",
       "      <td>CH1</td>\n",
       "      <td>n,652007..652311</td>\n",
       "      <td>306 nt</td>\n",
       "      <td>1</td>\n",
       "      <td>+1</td>\n",
       "      <td>-1</td>\n",
       "      <td></td>\n",
       "      <td>102 AA</td>\n",
       "      <td>102+0=102</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>XSETSPSIFPLSLGNNDPAGQVVIGCLVQGFFPSAPLSVTWNQNGD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KT723008</td>\n",
       "      <td>IGHA*01</td>\n",
       "      <td>Bos taurus_Holstein</td>\n",
       "      <td>F</td>\n",
       "      <td>H-CH2</td>\n",
       "      <td>g,652493..652821</td>\n",
       "      <td>330 nt</td>\n",
       "      <td>1</td>\n",
       "      <td>+1</td>\n",
       "      <td>-1</td>\n",
       "      <td></td>\n",
       "      <td>110 AA</td>\n",
       "      <td>110+0=110</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>DSSSCCVPNCEPSLSVQPPALEDLLLGSNASLTCTLSGLKSAEGAS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KT723008</td>\n",
       "      <td>IGHA*01</td>\n",
       "      <td>Bos taurus_Holstein</td>\n",
       "      <td>F</td>\n",
       "      <td>CH3-CHS</td>\n",
       "      <td>g,653011..653402</td>\n",
       "      <td>393 nt</td>\n",
       "      <td>1</td>\n",
       "      <td>+1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>131 AA</td>\n",
       "      <td>131+0=131</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>GNTFRPQVHLLPPPSEELALNELVTLTCLVRGFSPKEVLVRWLQGN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KT723008</td>\n",
       "      <td>IGHA*01</td>\n",
       "      <td>Bos taurus_Holstein</td>\n",
       "      <td>F</td>\n",
       "      <td>M</td>\n",
       "      <td>g,655709..655893</td>\n",
       "      <td>186 nt</td>\n",
       "      <td>1</td>\n",
       "      <td>+1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>62 AA</td>\n",
       "      <td>62+0=62</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>EHQPWLVLDLMQSSPEEDSPEASLWPTTVTLLTLFLLSLFYSTALT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KT723008</td>\n",
       "      <td>IGHD*02</td>\n",
       "      <td>Bos taurus_Holstein</td>\n",
       "      <td>ORF</td>\n",
       "      <td>CH1</td>\n",
       "      <td>n,525572..525894</td>\n",
       "      <td>324 nt</td>\n",
       "      <td>1</td>\n",
       "      <td>+1</td>\n",
       "      <td>-1</td>\n",
       "      <td></td>\n",
       "      <td>108 AA</td>\n",
       "      <td>108+0=108</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>XGESHLRVFSLVSCVSSPSDESTVALGCLARDFVPNSVSFSWKFNN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  accession number     name              species functionality  \\\n",
       "0         KT723008  IGHA*01  Bos taurus_Holstein             F   \n",
       "1         KT723008  IGHA*01  Bos taurus_Holstein             F   \n",
       "2         KT723008  IGHA*01  Bos taurus_Holstein             F   \n",
       "3         KT723008  IGHA*01  Bos taurus_Holstein             F   \n",
       "4         KT723008  IGHD*02  Bos taurus_Holstein           ORF   \n",
       "\n",
       "  exon/name/label start end positions number of nucleotides codon start  +n  \\\n",
       "0             CH1    n,652007..652311                306 nt           1  +1   \n",
       "1           H-CH2    g,652493..652821                330 nt           1  +1   \n",
       "2         CH3-CHS    g,653011..653402                393 nt           1  +1   \n",
       "3               M    g,655709..655893                186 nt           1  +1   \n",
       "4             CH1    n,525572..525894                324 nt           1  +1   \n",
       "\n",
       "  +/-n error corrections number of AA number of char partial  \\\n",
       "0   -1                         102 AA      102+0=102           \n",
       "1   -1                         110 AA      110+0=110           \n",
       "2                              131 AA      131+0=131           \n",
       "3                               62 AA        62+0=62           \n",
       "4   -1                         108 AA      108+0=108           \n",
       "\n",
       "  reverse complementary                                              FASTA  \n",
       "0                        XSETSPSIFPLSLGNNDPAGQVVIGCLVQGFFPSAPLSVTWNQNGD...  \n",
       "1                        DSSSCCVPNCEPSLSVQPPALEDLLLGSNASLTCTLSGLKSAEGAS...  \n",
       "2                        GNTFRPQVHLLPPPSEELALNELVTLTCLVRGFSPKEVLVRWLQGN...  \n",
       "3                        EHQPWLVLDLMQSSPEEDSPEASLWPTTVTLLTLFLLSLFYSTALT...  \n",
       "4                        XGESHLRVFSLVSCVSSPSDESTVALGCLARDFVPNSVSFSWKFNN...  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('IMGT_GENE.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dataprocess",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
